2-Draft-Accepted/RFCxxxx-DSC-for-PowerShell7.md

```diff
@@ -0,0 +1,323 @@
+
+About a year ago, we worked with
+[Gael Colas](https://twitter.com/gaelcolas)
+to author an RFC describing fundamental changes to
+[`Invoke-DscResource`](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/invoke-dscresource)
+in PowerShell 7.0.
+This enabled executing the Get, Set, and Test methods on DSC resources bypassing
+the LCM on Windows and removing the need for OMI on Linux. This was the start
+to enable cross platform DSC resources to be authored in PowerShell script (with
+some limitations).
+
+The changes to DSC shipped as an
+[experimental feature](https://docs.microsoft.com/powershell/scripting/learn/experimental-features)
+in PowerShell 7.0.
+
+### Bring your own agent (BYOA)
+
+A significant change from existing DSC for Windows PowerShell is there will be no
+default agent (LCM) shipped with PowerShell 7. Instead, we want to make it easier
+to integrate with existing configuration agents like
+[Ansible](https://docs.ansible.com/ansible/latest/user_guide/windows_dsc.html),
+[Chef](https://docs.chef.io/resources/dsc_resource/),
+[Puppet](https://forge.puppet.com/puppetlabs/dsc),
+[Azure Policy Guest Configuration](https://docs.microsoft.com/azure/governance/policy/concepts/guest-configuration),
+and other popular configuration solutions.
+
+Custom agents can use the existing PowerShell API (or call out to `pwsh`) to
+directly invoke DSC resources. For ad hoc configuration management, solutions can
+even use a PowerShell script run by TaskScheduler or cron.
+
+Along with the lack of a default agent, there is no pre-defined push or pull
+model for deploying DSC.
+
+### Move away from WMI/OMI/MOF
+
+DSC on Windows was originally designed relying on
+[WMI](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page)
+as a host on Windows,
+[OMI](https://www.dmtf.org/content/open-management-interface-omi-%E2%80%93-open-source-implementation-dmtf-cimwbem-standards)
+as a host on Linux, and
+[MOF](https://docs.microsoft.com/windows/win32/wmisdk/managed-object-format--mof-)
+for configuration and schema.
+
+MOF as a format is not something most developers are familiar with and there is no
+tooling support. However, developers of DSC resources written in PowerShell script
+still needed to learn MOF and also keep their MOF schema in sync with their script
+implementation.
+
+Today, when you execute a
+[DSC configuration script](https://docs.microsoft.com/powershell/scripting/dsc/configurations/configurations),
+the output is a MOF configuration file. In general, users do not need to read this
+MOF file as it's intended to be used by machines. However, MOF parsing creates a
+dependency on WMI/OMI. Instead, we will move towards a JSON configuration file format.
+
+Since there is no default agent to understand the JSON configuration file, it will
+be up to users or 3rd party agents to understand the configuration file.
+
+Moving from WMI/OMI also means no longer supporting C++ resources nor Python based resources.
```

Comment:
i assume that someone could build their own python based resources if they also build an agent for them.

Reply:
No, as use of DSC Resources is via the *-DscResource cmdlets, those only understand PowerShell classes, so unless they transcompile to a PowerShell class, then other languages are not currently planned to be supported